Solarwatt flex: improve soc detection

PeterFlorian opened the pull request:
It seems there are several variants of SOLARWATT-flex 1.0 manager, providing SoC on different endpoints. The template now supports `.*_batteryChannel_state_of_charge` and alternatively `.*_batteryChannelGroup_batteryStateOfCharge`.

PeterFlorian committed:
Second variant to determine SoC on Solarwatt-flex

It seems there are several variants of SOLARWATT-flex 1.0 manager, providing SoC on different endpoints.
The template now supports `.*_batteryChannel_state_of_charge` and alternatively
`.*_batteryChannelGroup_batteryStateOfCharge`.

andig commented on templates/definition/meter/solarwatt-flex.yaml:
```diff
@@ -1,14 +1,15 @@
 template: solarwatt-flex
+covers: ["solarwatt-myreserve"]
```
das Template gibts nicht- was ist hier bezweckt?

PeterFlorian replied:
Copy/Paste Fehler. -> behoben.

andig commented on templates/definition/meter/solarwatt-flex.yaml:
```diff
@@ -1,14 +1,15 @@
   description:
     en: |
-      Combines data of all connected PV inverters or batteries.
+      Template combines data of all connected PV inverters or batteries.
```
Dem Anwender ist egal, ob wir das "Template" nennen. Bitte rückgängig.

PeterFlorian replied:
Copy/Paste Fehler. -> behoben.

andig commented on templates/definition/meter/solarwatt-flex.yaml:
```diff
@@ -1,14 +1,15 @@
+      Template combines data of all connected PV inverters or batteries.
     de: |
-      Kombiniert Daten von allen verbundenen Solar-Wechselrichtern oder Batterien.
+      Template kombiniert Daten von allen verbundenen Solar-Wechselrichtern oder Batterien.
```
dito

PeterFlorian replied:
Copy/Paste Fehler. -> behoben.

andig commented on templates/definition/meter/solarwatt-flex.yaml:
```diff
@@ -22,7 +23,7 @@ render: |
     source: http
     uri: http://{{ .host }}/rest/items # EnergyManager flex
-    jq: >
+    jq: |
```
ist diese Änderung notwendig? falls nein bitte revert

PeterFlorian replied:
Bei SoC sind wegen der Kommentare Zeilenumbrüche erforderlich. Spricht etwas dagegen, es im ganzen File einheitlich zu halten?

andig commented on templates/definition/meter/solarwatt-flex.yaml:
```diff
@@ -66,13 +67,21 @@ render: |
+      # Variant 1
+      ([(.[] | select(.name | test(".*_batteryChannel_state_of_charge$")).state | parseFirstNumber)] | add // 0) as $soc
+      |
+      if $soc > 0 then
```
0 ist ja ein legaler soc- müsste man nicht eher schauen ob die erste expression irgendwas verarbeitet hat?

PeterFlorian replied:
Guter Punkt!
Mir ist auch noch aufgefallen, dass man Prozentwerte von mehreren Batterien nicht addieren darf. Dementsprechend ist nun die ganze Berechnung einfacher.

PeterFlorian committed:
Second variant to determine SoC on Solarwatt-flex

Fix PR comments

PeterFlorian committed:
Solarwatt-flex: No summation of percentage values for SoC

State of charge (SoC) values given in percent must not simply be added.

PeterFlorian committed:
Update templates/definition/meter/solarwatt-flex.yaml

Co-authored-by: premultiply <4681172+premultiply@users.noreply.github.com>

PeterFlorian commented:
Hallo @premultiply,
das wäre sicherlich sinnvoll.
Beim Solarwatt Flex Manager werden die angeschlossenen Geräte nicht über einen Index, sondern über eine ID angesprochen. Bei mir hat Solarwatt selbst die Ersteinrichtung vorgenommen und hier die Seriennummern verwendet. Ich kann nicht sagen ob das immer so ist. Auch ob die Speicher immer in der Reihenfolge der Registrierung gelistet werden kann ich nicht sagen.
Ich selbst habe nur einen Speicher, kann also auch nicht testen ob sich eine Änderung wie gewünscht verhält.
Unterm Strich sind mir das zu viele Unbekannte um das Feature so umzusetzen.
Die aktuelle Implementation sollte bei allen Anwendern funktionieren, die einen Speicher haben. bei mehreren Speichern muss man sich ohnehin stärker in die Konfiguration einarbeiten.

andig commented:
> Auch ob die Speicher immer in der Reihenfolge der Registrierung gelistet werden kann ich nicht sagen.

Haben die Speicher denn keine Id's? Kann man das da nicht analog machen?

PeterFlorian commented:
Der Solarwatt flex Manager scheint auf OpenHAB zu basieren. Zumindest kann ich unter http://SMF00-000002963.fritz.box/rest/items eine ganze Menge Daten abfragen. Der Name für den Ladezustand meines Speichers lautet "`vartabatteries`\_battery\_`130121607`\_batteryChannel\_state\_of\_charge", enthält also den Hersteller und die Seriennummer. Vermutlich ist das immer so. Ich kann es aber nicht mit Sicherheit sagen, da ich es nicht eingerichtet und keinen Zugang für den Manager habe. Vielleicht war das auch nur ein guter Gedanke des Installateurs.
Selbst wenn dieses Schema immer so ist halte ich es für schwierig, dass die Nutzer des Templates die Daten korrekt eingeben. Ich selbst hätte wahrscheinlich als Hersteller eher "VARTA" angegeben als "vartabatteries". Um dem zu Begegnen müsste man also eine Liste aller Hersteller hinterlegen, nur um die Schreibweise sicherzustellen. Keine Ahnung ob das in der Template-Sprache geht. Und wer pflegt dann diese Liste? Ich kenne ja nur "vartabatteries".
Das Template kann ja nur einen Speicher auslesen. Das ist glaube ich bei allen Templates so. Wenn also ein Nutzer einen zweiten Speicher hat, muss er sich ohnehin mit der Konfiguration des zweiten Speichers beschäftigen. Entweder findet er heraus, welcher Speicher schon über den Manager konfiguriert ist oder er ersetzt die SoC-Konfiguration komplett.

premultiply commented on templates/definition/meter/solarwatt-flex.yaml:
```diff
@@ -66,13 +68,20 @@ render: |
+      first(
+        # Variant 1
+        (.[] | select(.name | test(".*_batteryChannel_state_of_charge$")).state | parseFirstNumber)
+        //
+        # Variant 2
+        (.[] | select(.name | test(".*_batteryChannelGroup_batteryStateOfCharge$")).state | tonumber)
+      )
```
ping @PeterFlorian

Funktioniert das so? Getestet?

andig replied:
Einfach Copilot ignorieren. Es zählt nur obs funktioniert und getestet ist.

PeterFlorian replied:
Hallo @andig ,
ich hatte es vor der Änderung für den Copilot positiv getestet und hinterher.
Ob die Ergänzung in Zeile 77 funktioniert kann ich nicht sicher sagen, da ich eine andere Version des EnergyManagers habe. Nach Releasen will ich @alb-sch bitten mal zu testen (siehe https://github.com/evcc-io/evcc/discussions/19204).

PeterFlorian committed:
Solarwatt-flex: No summation of percentage values for SoC

Implement (and fix!) suggestion of GitHub Copilot.
Suggestion mixed up jq phrase ""//" with a comment, wich has to be started by a hash.

PeterFlorian committed:
Merge remote-tracking branch 'origin/feature/solarwatt-flex_variant_for_SoC' into feature/solarwatt-flex_variant_for_SoC

# Conflicts:
#	templates/definition/meter/solarwatt-flex.yaml

premultiply commented:
Wichtig wäre es vor dem Merge zu testen, da es sonst im dümmsten Fall ein funktionierendes System kaputt macht.

PeterFlorian commented:
Es geht definitiv nichts kaputt. Ich selbst hatte das Template damals für meine Anlage geschrieben. Auch nach der Änderung funktioniert diese einwandfrei. Die Ergänzung habe ich mit generierten Testdaten überprüft, muss mich dabei aber auf die Angaben aus https://github.com/evcc-io/evcc/discussions/19204 verlassen.
Einen echten Test kann ich ohne die Hardware nicht ausführen. Das müsste also @alb-sch erledigen. Ich befürchte, dass er keine Entwicklungsumgebung aufbauen möchte, um den Test durchzuführen, zumal er inzwischen eine Lösung ohne Template im Einsatz hat. Oder gibt es eine andere Möglichkeit, wie er testen kann ohne evcc selbst zu compilieren?

PeterFlorian committed:
Merge branch 'master' into feature/solarwatt-flex_variant_for_SoC

PeterFlorian committed:
Merge branch 'master' into feature/solarwatt-flex_variant_for_SoC

andig commented:
> Oder gibt es eine andere Möglichkeit, wie er testen kann ohne evcc selbst zu compilieren?

Ja, die gibts schon länger: https://github.com/evcc-io/evcc/pull/19495

andig commented:
Wir machen mal mangels Fortschritt zu. Wenns ein Testergebnis gibt gerne rein damit.

xuckz commented:
Mich betrifft das Problem auch, und ich würde mich über einen Fix freuen.

Aktuell starte ich das gefixte Template im Docker Image mit "evcc --template-type ..." und kann die Funktion der Batterie als auch der PV  bestätigen.
